# Data Frames with Pandas

## Acknowledgment 

Some of the content here is based on [Computational and Inferential Thinking: The Foundations of Data Science](https://inferentialthinking.com/chapters/intro.html), by A. Adhikari, J. DeNero, D. Wagner.

On the other hand, this text uses its own module `datascience` for data frame manipulations, while we will use pandas, which is the most commonly used library for data frames in Python.

## Introduction

We call data organized in tables *data frames*.

It can be viewed in two ways:

* a sequence of named columns, each describing a single aspect of all entries in a data set, or
* a sequence of rows, each containing all information about a single entry in a data set.

## Pandas

Our main tool for dealing with data frames will be [pandas](https://pandas.pydata.org/), which is the most popular library for dealing with data frames in Python.

Pandas uses NumPy internally to store the data and perform the operations with rows and columns, and is therefore fast enough to deal with large amounts of data.  (Although there are alternatives better optimized for *huge* amounts of data.)

### Installing Pandas

pandas does not come with Python, so it needs to be installed separately.  If you have a *vanilla* installation of Python, you can do it by running 

```
pip install pandas
```

from a terminal.

On the other hand, if you installed Anaconda, it should already be available.

### Loading Pandas

As usual, we need to first import pandas so that we can use its features.  We usually give the (standard) shortcut `pd`:

In [88]:
import pandas as pd

(Thus we can call its functions with `pd.name_of_the_function` instead of `pandas.name_of_the_function`.)

We also usually need arrays when dealing with  data frames, so let's import NumPy as well:

In [89]:
import numpy as np

## Creating Data Frames

### Creating Data Frames Manually

We can use *dictionaries* to create a data frames.  The keys are the *labels* for each column (usually a string), and the values are the arrays (or lists) of values for each column.

The syntax is
```python
pd.DataFrame({ 
    "name of col 1": col_1,
    "name of col 2": col_2,
    ...
    "name of col n": col_n,
    
  })
```

For example, let's create an array containing information about flowers:

In [90]:
flowers = pd.DataFrame({
    "Name": np.array(["lotus", "sunflower", "rose"]),
    "Number of petals": np.array([8, 34, 5])

})

flowers

Name  Number of petals
0      lotus                 8
1  sunflower                34
2       rose                 5

Note that we automatically get numbers on the left of the rows.  These numbers form the *index* of the data frame, which is used to identify the rows.  As usual, it starts with zero.  Later we will see how we can make some column the index of the data frame.

### Reading from Files

Usually we read data from files.  The most common type of file used to store data are [comma-separated values](https://en.wikipedia.org/wiki/Comma-separated_values) files, usually referred to as CSV files.

To get the content of a CSV into a data frame, we use the function `pd.read_csv`, with the name of the file passed as a *string*.  The file has to be in the same folder as our notebook, or the the path to it has to be given in the string.

As an example, let's load the file [nba_salaries.csv](nba_salaries.csv) (provided with this notebook), which contain the [salaries of all National Basketball Association players](https://www.statcrunch.com/app/index.php?dataid=1843341) in 2015-2016, and save the resulting data frame as `nba`:

In [91]:
nba = pd.read_csv("nba_salaries.csv")
nba

PLAYER POSITION                TEAM  '15-'16 SALARY
0       Paul Millsap       PF       Atlanta Hawks       18.671659
1         Al Horford        C       Atlanta Hawks       12.000000
2     Tiago Splitter        C       Atlanta Hawks        9.756250
3        Jeff Teague       PG       Atlanta Hawks        8.000000
4        Kyle Korver       SG       Atlanta Hawks        5.746479
..               ...      ...                 ...             ...
412        Gary Neal       PG  Washington Wizards        2.139000
413     DeJuan Blair        C  Washington Wizards        2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards        1.920240
415   Garrett Temple       SG  Washington Wizards        1.100602
416     Jarell Eddie       SG  Washington Wizards        0.561716

[417 rows x 4 columns]

Each row represents one player. The columns are:

| **Column Label**   | Description                                         |
|--------------------|-----------------------------------------------------|
| `PLAYER`           | Player's name                                       |
| `POSITION`         | Player's position on team                           |
| `TEAM`             | Team name                                           |
|`'15-'16 SALARY`            | Player's salary in 2015-2016, in millions of dollars|
 
The code for the positions is PG (Point Guard), SG (Shooting Guard), PF (Power Forward), SF (Small Forward), and C (Center).

The first row shows that Paul Millsap, Power Forward for the Atlanta Hawks, had a salary of almost $\$18.7$ million in 2015-2016.

`pd.read_csv` has many options, which allow you to skip rows at the beginning, choose columns, choose the index, etc.  You can [read the documentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) or call

```python
help(pd.read_csv)
```

Pandas can also read from other formats, including Microsoft Excel (with `pd.read_excel`), but here will only use CSV files.

### Reading from URL

We can also use `pd.read_csv` to read a CSV available from a URL.  For instance,

[http://www2.census.gov/programs-surveys/popest/technical-documentation/file-layouts/2010-2019/nc-est2019-agesex-res.csv](http://www2.census.gov/programs-surveys/popest/technical-documentation/file-layouts/2010-2019/nc-est2019-agesex-res.csv)

is a link for a CSV file containing some US Census data.  As long as the link is valid (which is, as of July 2023), passing the URL as a *string* to `pd.read_csv` should load the data frame:

In [92]:
data_url = "http://www2.census.gov/programs-surveys/popest/technical-documentation/file-layouts/2010-2019/nc-est2019-agesex-res.csv"

census_df = pd.read_csv(data_url)

census_df

SEX  AGE  CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  \
0      0    0        3944153            3944160          3951430   
1      0    1        3978070            3978090          3957730   
2      0    2        4096929            4096939          4090621   
3      0    3        4119040            4119051          4111688   
4      0    4        4063170            4063186          4077346   
..   ...  ...            ...                ...              ...   
301    2   97          53582              53605            54165   
302    2   98          36641              36675            37567   
303    2   99          26193              26214            26098   
304    2  100          44202              44246            45080   
305    2  999      156964212          156969328        157246908   

     POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  POPESTIMATE2014  \
0            3963092          3926570          3931258          3954787   
1            3966225          3977549          3942698          3948891   
2            3970654          3978925          3991740          3958711   
3            4101644          3981531          3991017          4005928   
4            4121488          4111490          3992502          4004032   
..               ...              ...              ...              ...   
301            57123            59221            60670            62407   
302            40078            42561            43866            45887   
303            26995            29076            30672            32245   
304            47427            50040            52817            57108   
305        158356816        159456593        160511789        161646584   

     POPESTIMATE2015  POPESTIMATE2016  POPESTIMATE2017  POPESTIMATE2018  \
0            3983981          3954773          3893990          3815343   
1            3973133          4002903          3972711          3908830   
2            3966321          3991349          4020045          3987032   
3            3974351          3982984          4006946          4033038   
4            4020292          3989750          3997280          4018719   
..               ...              ...              ...              ...   
301            68608            69709            76276            82100   
302            46611            52122            52978            57593   
303            33560            34278            38489            39576   
304            60309            64381            67113            72153   
305        162778769        163919526        164939619        165801767   

     POPESTIMATE2019  
0            3783052  
1            3829599  
2            3922044  
3            3998665  
4            4043323  
..               ...  
301            84358  
302            62607  
303            42528  
304            76850  
305        166582199  

[306 rows x 14 columns]

## Renaming Columns

Let's relabel the last column simply `SALARY`:

In [93]:
nba = nba.rename(columns={"'15-'16 SALARY": "SALARY"})  # overwrite nba data frame
nba

PLAYER POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

The syntax to rename various columns is:
```python
data_frame_name.rename(
    columns={
        old_col_name_1: new_col_name_1,
        old_col_name_2: new_col_name_2,
        ...
        old_col_name_last: new_col_name_last,
    } )
```

In other words, we pass to the method `rename` the argument `columns=` (or it will try to rename the index, by default) with a dictionary whose the key/value pairs contain the old names as keys and the new names as values.

## Data Frame Properties

The attribute `shape` gives a pair: the number of rows and number of columns (in this order):

In [94]:
nba.shape

(417, 4)

Note that `shape` is an attribute and not a method/function, and therefore *should not* be followed by parentheses.

We can also get the number of rows with `len`, as usual:

In [95]:
len(nba)

417

To get the column names, we can use the attribute `columns`:

In [96]:
nba.columns

Index(['PLAYER', 'POSITION', 'TEAM', 'SALARY'], dtype='object')

To get the index of the data frame (which is not very interesting in this case):

In [97]:
nba.index

RangeIndex(start=0, stop=417, step=1)

We can also see the data types for each column:

In [98]:
nba.dtypes

PLAYER       object
POSITION     object
TEAM         object
SALARY      float64
dtype: object

(Note that `object` is basically a string.)

We can get the top rows of a data frame using the `head` method.  For instance, to get the top 10 rows:

In [99]:
nba.head(10)

PLAYER POSITION           TEAM     SALARY
0      Paul Millsap       PF  Atlanta Hawks  18.671659
1        Al Horford        C  Atlanta Hawks  12.000000
2    Tiago Splitter        C  Atlanta Hawks   9.756250
3       Jeff Teague       PG  Atlanta Hawks   8.000000
4       Kyle Korver       SG  Atlanta Hawks   5.746479
5   Thabo Sefolosha       SF  Atlanta Hawks   4.000000
6        Mike Scott       PF  Atlanta Hawks   3.333333
7     Kent Bazemore       SF  Atlanta Hawks   2.000000
8   Dennis Schroder       PG  Atlanta Hawks   1.763400
9  Tim Hardaway Jr.       SG  Atlanta Hawks   1.304520

We can use `tail` for the bottom rows.  So, to get the bottom 5 rows:

In [100]:
nba.tail(5)

PLAYER POSITION                TEAM    SALARY
412        Gary Neal       PG  Washington Wizards  2.139000
413     DeJuan Blair        C  Washington Wizards  2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards  1.920240
415   Garrett Temple       SG  Washington Wizards  1.100602
416     Jarell Eddie       SG  Washington Wizards  0.561716

## Changing the Index

In this case, it might make sense to make the column of players the *index*, as each row has data about a particular player:

In [101]:
nba_by_player = nba.set_index("PLAYER")

nba_by_player

POSITION                TEAM     SALARY
PLAYER                                                 
Paul Millsap          PF       Atlanta Hawks  18.671659
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Jeff Teague           PG       Atlanta Hawks   8.000000
Kyle Korver           SG       Atlanta Hawks   5.746479
...                  ...                 ...        ...
Gary Neal             PG  Washington Wizards   2.139000
DeJuan Blair           C  Washington Wizards   2.000000
Kelly Oubre Jr.       SF  Washington Wizards   1.920240
Garrett Temple        SG  Washington Wizards   1.100602
Jarell Eddie          SG  Washington Wizards   0.561716

[417 rows x 3 columns]

The old index is gone and now our rows are identified by the corresponding player's name.

## Selecting Columns

We select a column of a data frame by simply passing its label.  For instance, we can select the `SALARY` column of `nba_by_player`:

In [102]:
nba_by_player["SALARY"]

PLAYER
Paul Millsap       18.671659
Al Horford         12.000000
Tiago Splitter      9.756250
Jeff Teague         8.000000
Kyle Korver         5.746479
                     ...    
Gary Neal           2.139000
DeJuan Blair        2.000000
Kelly Oubre Jr.     1.920240
Garrett Temple      1.100602
Jarell Eddie        0.561716
Name: SALARY, Length: 417, dtype: float64

Note that the result is *not a data frame*, but a [series](https://pandas.pydata.org/docs/reference/api/pandas.Series.html), which is basically a NumPy array with some extra metadata, such as name (SALARY in this case) and an index.  But, since it is basically a NumPy array, we can use most of NumPy's functions with series.  For instance, if we want to know the average salary in the data frame:

In [103]:
np.mean(nba_by_player["SALARY"])

5.074814482014388

If we need a true NumPy array, we can use `.to_numpy()`:

In [104]:
nba_by_player["SALARY"].to_numpy()

array([18.671659, 12.      ,  9.75625 ,  8.      ,  5.746479,  4.      ,
        3.333333,  2.      ,  1.7634  ,  1.30452 ,  1.      ,  0.947276,
        0.525093,  0.525093,  7.730337,  6.912869,  6.796117,  5.      ,
        3.43104 ,  3.42551 ,  2.616975,  2.56926 ,  2.16516 ,  2.038206,
        1.82436 ,  1.74984 ,  1.706225,  1.17096 ,  1.14864 , 24.894863,
       11.235955,  6.3     ,  3.42551 ,  1.59984 ,  1.5     ,  1.5     ,
        1.362897,  1.33548 ,  1.185784,  1.14024 ,  0.981348,  0.845059,
       13.5     , 13.125306, 12.      ,  7.      ,  6.331404,  6.110034,
        5.675   ,  4.2042  ,  3.034356,  2.85494 ,  2.61252 ,  2.139   ,
        1.20144 ,  0.947276,  0.947276,  0.947276,  0.947276,  0.525093,
       20.093064, 16.4075  , 13.4     ,  7.44876 ,  5.543725,  4.5     ,
        2.85494 ,  2.38044 ,  2.25    ,  1.53588 ,  1.39116 ,  1.015421,
       22.9705  , 19.689   , 16.407501, 14.26087 , 10.5225  ,  8.988765,
        4.95    ,  2.1     ,  1.276   ,  1.147276, 

On the other hand, the series has its own methods, so we could have found the average salary with:

In [105]:
nba_by_player["SALARY"].mean()

5.074814482014388

Many of the functions provided by NumPy for arrays are available for series as methods, which is the preferred way to use it.

An alternative is to get columns from the corresponding attributes.  For instance, the following is equivalent to `nba_by_player["SALARY"]`:

In [106]:
nba_by_player.SALARY

PLAYER
Paul Millsap       18.671659
Al Horford         12.000000
Tiago Splitter      9.756250
Jeff Teague         8.000000
Kyle Korver         5.746479
                     ...    
Gary Neal           2.139000
DeJuan Blair        2.000000
Kelly Oubre Jr.     1.920240
Garrett Temple      1.100602
Jarell Eddie        0.561716
Name: SALARY, Length: 417, dtype: float64

The problem with this second approach is that the column label cannot have a space.  Due to this limitation, we will always use the first approach (with the brackets `[ ]`).

We can also select a column by position (*counting from 0*, as usual) using `.iloc`.  For example, to get the second column (TEAM) we can do:

In [107]:
nba_by_player.iloc[:, 1]

PLAYER
Paul Millsap            Atlanta Hawks
Al Horford              Atlanta Hawks
Tiago Splitter          Atlanta Hawks
Jeff Teague             Atlanta Hawks
Kyle Korver             Atlanta Hawks
                          ...        
Gary Neal          Washington Wizards
DeJuan Blair       Washington Wizards
Kelly Oubre Jr.    Washington Wizards
Garrett Temple     Washington Wizards
Jarell Eddie       Washington Wizards
Name: TEAM, Length: 417, dtype: object

**Note that `.iloc` is followed by square brackets, not parentheses!**

`.iloc` selects *rows* and columns and we can use *slicing* to select them.  Thus, the `:` in `.iloc[:, 2]` told it to get *all* rows.  (Rows are first, then the columns.)

We can also select more than one column by passing a list of labels inside the square brackets:

In [108]:
nba_by_player[["SALARY", "TEAM"]]  # note the double brackets [[ ]]

SALARY                TEAM
PLAYER                                        
Paul Millsap     18.671659       Atlanta Hawks
Al Horford       12.000000       Atlanta Hawks
Tiago Splitter    9.756250       Atlanta Hawks
Jeff Teague       8.000000       Atlanta Hawks
Kyle Korver       5.746479       Atlanta Hawks
...                    ...                 ...
Gary Neal         2.139000  Washington Wizards
DeJuan Blair      2.000000  Washington Wizards
Kelly Oubre Jr.   1.920240  Washington Wizards
Garrett Temple    1.100602  Washington Wizards
Jarell Eddie      0.561716  Washington Wizards

[417 rows x 2 columns]

Note that we obtain a data frame, the index is preserved, and the order in the given list is used.

In particular, if we want a single row as a data frame, and not a series, we can use the double square brackets again:

In [109]:
nba[["SALARY"]]  # using nba, not nba_by_player

SALARY
0    18.671659
1    12.000000
2     9.756250
3     8.000000
4     5.746479
..         ...
412   2.139000
413   2.000000
414   1.920240
415   1.100602
416   0.561716

[417 rows x 1 columns]

We can also use `iloc` to select more than a single columns.  For instance, the last two columns of `nba`:

In [110]:
nba.iloc[:, -2:]

TEAM     SALARY
0         Atlanta Hawks  18.671659
1         Atlanta Hawks  12.000000
2         Atlanta Hawks   9.756250
3         Atlanta Hawks   8.000000
4         Atlanta Hawks   5.746479
..                  ...        ...
412  Washington Wizards   2.139000
413  Washington Wizards   2.000000
414  Washington Wizards   1.920240
415  Washington Wizards   1.100602
416  Washington Wizards   0.561716

[417 rows x 2 columns]

Or the first and third, using a list of column indices:

In [111]:
nba.iloc[:, [0, 2]]

PLAYER                TEAM
0       Paul Millsap       Atlanta Hawks
1         Al Horford       Atlanta Hawks
2     Tiago Splitter       Atlanta Hawks
3        Jeff Teague       Atlanta Hawks
4        Kyle Korver       Atlanta Hawks
..               ...                 ...
412        Gary Neal  Washington Wizards
413     DeJuan Blair  Washington Wizards
414  Kelly Oubre Jr.  Washington Wizards
415   Garrett Temple  Washington Wizards
416     Jarell Eddie  Washington Wizards

[417 rows x 2 columns]

The index (which is not quite a column) is given by the `.index` attribute (no parentheses!):

In [112]:
nba_by_player.index

Index(['Paul Millsap', 'Al Horford', 'Tiago Splitter', 'Jeff Teague',
       'Kyle Korver', 'Thabo Sefolosha', 'Mike Scott', 'Kent Bazemore',
       'Dennis Schroder', 'Tim Hardaway Jr.',
       ...
       'Kris Humphries', 'Jared Dudley', 'Alan Anderson', 'Drew Gooden',
       'Ramon Sessions', 'Gary Neal', 'DeJuan Blair', 'Kelly Oubre Jr.',
       'Garrett Temple', 'Jarell Eddie'],
      dtype='object', name='PLAYER', length=417)

It's not a series, since it has no index itself, but it is also an array with extra metadata:

In [113]:
nba_by_player.index.to_numpy()

array(['Paul Millsap', 'Al Horford', 'Tiago Splitter', 'Jeff Teague',
       'Kyle Korver', 'Thabo Sefolosha', 'Mike Scott', 'Kent Bazemore',
       'Dennis Schroder', 'Tim Hardaway Jr.', 'Walter Tavares',
       'Jason Richardson', 'Lamar Patterson', 'Terran Petteway',
       'Avery Bradley', 'Isaiah Thomas', 'Jae Crowder', 'Jonas Jerebko',
       'Marcus Smart', 'Evan Turner', 'Tyler Zeller', 'Jared Sullinger',
       'Kelly Olynyk', 'Perry Jones', 'Terry Rozier', 'James Young',
       'Zoran Dragic', 'Jordan Mickey', 'R.J. Hunter', 'Joe Johnson',
       'Thaddeus Young', 'Jarrett Jack', 'Bojan Bogdanovic',
       'Sergey Karasev', 'Wayne Ellington', 'Shane Larkin',
       'Andrea Bargnani', 'Rondae Hollis-Jefferson', 'Earl Clark',
       'Chris McCullough', 'Thomas Robinson', 'Markel Brown',
       'Al Jefferson', 'Nicolas Batum', 'Kemba Walker', 'Marvin Williams',
       'Michael Kidd-Gilchrist', 'Spencer Hawes', 'Courtney Lee',
       'Cody Zeller', 'Jeremy Lamb', 'Brian Roberts',

## Selecting Rows

Suppose we want to get the information about a particular players, say Kobe Bryant.

In `nba_by_player`, since the players make the index, we can use `loc`:

In [114]:
nba_by_player.loc["Kobe Bryant"]

POSITION                    SF
TEAM        Los Angeles Lakers
SALARY                    25.0
Name: Kobe Bryant, dtype: object

This gives all the item in Kobe Bryant's row.

**Note:** `loc` filters rows and columns by *labels*, while `iloc` filters by *numerical index*.

But this would fail with `nba`, in which case the index is numeric, and, a priori, we do not know the number for his row.  (If we did, we could use `iloc`.)  In this case, we can  *filter* for the row.

Similar to NumPy's arrays, we can select rows of a data frame by passing to `loc` an array of booleans.  Thus we can do:

In [115]:
nba.loc[nba["PLAYER"] == "Kobe Bryant"]

PLAYER POSITION                TEAM  SALARY
169  Kobe Bryant       SF  Los Angeles Lakers    25.0

The result is different (the former was a series, and now we have a data frame), but contains the same information.

In either case, we can get, say, Kobe Bryant's salary:

In [116]:
nba_by_player.loc["Kobe Bryant"]["SALARY"]  # get SALARY from series

25.0

In [117]:
nba_by_player.loc["Kobe Bryant", "SALARY"]  # index, and column label

25.0

In [118]:
nba.loc[nba["PLAYER"] == "Kobe Bryant"]["SALARY"]

169    25.0
Name: SALARY, dtype: float64

Note that this last example, the salary comes as a series (since we had a data frame before), so we get the index with the salary as well.  This is the case because the filtering could result in more than one row.

Again, knowing that the index in `nba` for Kobe Bryant is 169, we could get his row (as a series) with:

In [119]:
nba.loc[169]

PLAYER             Kobe Bryant
POSITION                    SF
TEAM        Los Angeles Lakers
SALARY                    25.0
Name: 169, dtype: object

Note that `iloc` also works, since the index matches the position:

In [120]:
nba.iloc[169]

PLAYER             Kobe Bryant
POSITION                    SF
TEAM        Los Angeles Lakers
SALARY                    25.0
Name: 169, dtype: object

With `iloc` we can get slices of rows:

In [121]:
nba_by_player.iloc[10:20]  # rows 10 to 19

POSITION            TEAM    SALARY
PLAYER                                             
Walter Tavares          C   Atlanta Hawks  1.000000
Jason Richardson       SG   Atlanta Hawks  0.947276
Lamar Patterson        SG   Atlanta Hawks  0.525093
Terran Petteway        SG   Atlanta Hawks  0.525093
Avery Bradley          PG  Boston Celtics  7.730337
Isaiah Thomas          PG  Boston Celtics  6.912869
Jae Crowder            SF  Boston Celtics  6.796117
Jonas Jerebko          PF  Boston Celtics  5.000000
Marcus Smart           PG  Boston Celtics  3.431040
Evan Turner            SG  Boston Celtics  3.425510

And, as seen above, we can also select columns:

In [122]:
nba_by_player.iloc[10:20, -2:]  # rows 10 to 19, last two columns

TEAM    SALARY
PLAYER                                    
Walter Tavares     Atlanta Hawks  1.000000
Jason Richardson   Atlanta Hawks  0.947276
Lamar Patterson    Atlanta Hawks  0.525093
Terran Petteway    Atlanta Hawks  0.525093
Avery Bradley     Boston Celtics  7.730337
Isaiah Thomas     Boston Celtics  6.912869
Jae Crowder       Boston Celtics  6.796117
Jonas Jerebko     Boston Celtics  5.000000
Marcus Smart      Boston Celtics  3.431040
Evan Turner       Boston Celtics  3.425510

## Sorting

We can easily sort data frames with the method `sort_values`.  For instance, to sort by SALARY:

In [123]:
nba_by_player.sort_values("SALARY")

POSITION                 TEAM     SALARY
PLAYER                                                         
Thanasis Antetokounmpo       SF      New York Knicks   0.030888
Cory Jefferson               PF         Phoenix Suns   0.049709
Jordan McRae                 SG         Phoenix Suns   0.049709
Orlando Johnson              SG         Phoenix Suns   0.055722
Phil Pressey                 PG         Phoenix Suns   0.055722
...                         ...                  ...        ...
Dwight Howard                 C      Houston Rockets  22.359364
Carmelo Anthony              SF      New York Knicks  22.875000
LeBron James                 SF  Cleveland Cavaliers  22.970500
Joe Johnson                  SF        Brooklyn Nets  24.894863
Kobe Bryant                  SF   Los Angeles Lakers  25.000000

[417 rows x 3 columns]

(Some of the salaries that seem (relatively) quite small are for the player who changed teams mid season, and the data frame only keep their salaries from one team.)

**Note:** The operation above *returns a new data frame*, and does not alter the original.  To do so, you can pass the optional argument `inplace=True`, as in

```python
nba_by_player.sort_values("SALARY", inplace=True)
```

or simply overwrite the original with

```python
nba_by_player = nba_by_player.sort_values("SALARY")
```

***Important:* The same is true for most methods we will introduce here!  They output a new data frame, and do not change the original!**  Some of them will also have the `inplace` optional argument, but you can always just overwrite the original value, as done above, if necessary.

We can pass the optional argument `ascending=False`, to sort it in decreasing order:

In [124]:
nba_by_player.sort_values("SALARY", ascending=False)

POSITION                 TEAM     SALARY
PLAYER                                                         
Kobe Bryant                  SF   Los Angeles Lakers  25.000000
Joe Johnson                  SF        Brooklyn Nets  24.894863
LeBron James                 SF  Cleveland Cavaliers  22.970500
Carmelo Anthony              SF      New York Knicks  22.875000
Dwight Howard                 C      Houston Rockets  22.359364
...                         ...                  ...        ...
Elliot Williams              SG    Memphis Grizzlies   0.055722
Orlando Johnson              SG         Phoenix Suns   0.055722
Cory Jefferson               PF         Phoenix Suns   0.049709
Jordan McRae                 SG         Phoenix Suns   0.049709
Thanasis Antetokounmpo       SF      New York Knicks   0.030888

[417 rows x 3 columns]

We can also sort by the index:

In [125]:
nba_by_player.sort_index()

POSITION                    TEAM     SALARY
PLAYER                                                     
Aaron Brooks          PG           Chicago Bulls   2.250000
Aaron Gordon          PF           Orlando Magic   4.171680
Aaron Harrison        SG       Charlotte Hornets   0.525093
Adreian Payne         PF  Minnesota Timberwolves   1.938840
Al Horford             C           Atlanta Hawks  12.000000
...                  ...                     ...        ...
Wilson Chandler       SF          Denver Nuggets  10.449438
Zach LaVine           PG  Minnesota Timberwolves   2.148360
Zach Randolph         PF       Memphis Grizzlies   9.638555
Zaza Pachulia          C        Dallas Mavericks   5.200000
Zoran Dragic          SG          Boston Celtics   1.706225

[417 rows x 3 columns]

We can sort by more than one column, by passing a list of labels:

In [126]:
nba_by_player.sort_values(["POSITION", "SALARY"])

POSITION                   TEAM     SALARY
PLAYER                                                     
Ryan Hollins            C      Memphis Grizzlies   0.111444
Kendrick Perkins        C   New Orleans Pelicans   0.947276
Jeff Withey             C              Utah Jazz   0.947276
Robert Sacre            C     Los Angeles Lakers   0.981348
Walter Tavares          C          Atlanta Hawks   1.000000
...                   ...                    ...        ...
Klay Thompson          SG  Golden State Warriors  15.501000
Eric Gordon            SG   New Orleans Pelicans  15.514031
James Harden           SG        Houston Rockets  15.756438
Jimmy Butler           SG          Chicago Bulls  16.407500
Dwyane Wade            SG             Miami Heat  20.000000

[417 rows x 3 columns]

The data is sorted according to the first label, and rows that have the same value for this column are sorted according to the second label.

## Resetting the Index

Sometimes when we alter a data frame, a numerical index can get mixed.  For instance, when sorting:

In [127]:
nba.sort_values("POSITION")

PLAYER POSITION                    TEAM     SALARY
371  Jonas Valanciunas        C         Toronto Raptors   4.660482
75    Tristan Thompson        C     Cleveland Cavaliers  14.260870
227     Nikola Pekovic        C  Minnesota Timberwolves  12.100000
374    Bismack Biyombo        C         Toronto Raptors   2.814000
131      Dwight Howard        C         Houston Rockets  22.359364
..                 ...      ...                     ...        ...
166      Lester Hudson       SG    Los Angeles Clippers   1.015421
318     Archie Goodwin       SG            Phoenix Suns   1.160160
316       Devin Booker       SG            Phoenix Suns   2.127840
360      Manu Ginobili       SG       San Antonio Spurs   2.814000
416       Jarell Eddie       SG      Washington Wizards   0.561716

[417 rows x 4 columns]

In these cases often we prefer to *reset the index*:

In [128]:
nba.sort_values("POSITION").reset_index()

index             PLAYER POSITION                    TEAM     SALARY
0      371  Jonas Valanciunas        C         Toronto Raptors   4.660482
1       75   Tristan Thompson        C     Cleveland Cavaliers  14.260870
2      227     Nikola Pekovic        C  Minnesota Timberwolves  12.100000
3      374    Bismack Biyombo        C         Toronto Raptors   2.814000
4      131      Dwight Howard        C         Houston Rockets  22.359364
..     ...                ...      ...                     ...        ...
412    166      Lester Hudson       SG    Los Angeles Clippers   1.015421
413    318     Archie Goodwin       SG            Phoenix Suns   1.160160
414    316       Devin Booker       SG            Phoenix Suns   2.127840
415    360      Manu Ginobili       SG       San Antonio Spurs   2.814000
416    416       Jarell Eddie       SG      Washington Wizards   0.561716

[417 rows x 5 columns]

A new numerical index is given (in the new order), and the old is added a new column, labeled `index` by default.  A name can be given by passing the `names=` argument:

In [129]:
nba.sort_values("POSITION").reset_index(names="Old Index")

Old Index             PLAYER POSITION                    TEAM     SALARY
0          371  Jonas Valanciunas        C         Toronto Raptors   4.660482
1           75   Tristan Thompson        C     Cleveland Cavaliers  14.260870
2          227     Nikola Pekovic        C  Minnesota Timberwolves  12.100000
3          374    Bismack Biyombo        C         Toronto Raptors   2.814000
4          131      Dwight Howard        C         Houston Rockets  22.359364
..         ...                ...      ...                     ...        ...
412        166      Lester Hudson       SG    Los Angeles Clippers   1.015421
413        318     Archie Goodwin       SG            Phoenix Suns   1.160160
414        316       Devin Booker       SG            Phoenix Suns   2.127840
415        360      Manu Ginobili       SG       San Antonio Spurs   2.814000
416        416       Jarell Eddie       SG      Washington Wizards   0.561716

[417 rows x 5 columns]

If we do not want to keep the old index, we can use the optional argument `drop=True`:

In [130]:
nba.sort_values("POSITION").reset_index(drop=True)

PLAYER POSITION                    TEAM     SALARY
0    Jonas Valanciunas        C         Toronto Raptors   4.660482
1     Tristan Thompson        C     Cleveland Cavaliers  14.260870
2       Nikola Pekovic        C  Minnesota Timberwolves  12.100000
3      Bismack Biyombo        C         Toronto Raptors   2.814000
4        Dwight Howard        C         Houston Rockets  22.359364
..                 ...      ...                     ...        ...
412      Lester Hudson       SG    Los Angeles Clippers   1.015421
413     Archie Goodwin       SG            Phoenix Suns   1.160160
414       Devin Booker       SG            Phoenix Suns   2.127840
415      Manu Ginobili       SG       San Antonio Spurs   2.814000
416       Jarell Eddie       SG      Washington Wizards   0.561716

[417 rows x 4 columns]

## Filtering Rows

### Filtering with `loc`

As we've seen, we can use `loc` to filter rows by conditions.

For example, to find all rows for centers:

In [131]:
nba_by_player.loc[nba_by_player["POSITION"] == "C"]

POSITION                TEAM     SALARY
PLAYER                                                 
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Walter Tavares         C       Atlanta Hawks   1.000000
Tyler Zeller           C      Boston Celtics   2.616975
Jared Sullinger        C      Boston Celtics   2.569260
...                  ...                 ...        ...
Rudy Gobert            C           Utah Jazz   1.175880
Jeff Withey            C           Utah Jazz   0.947276
Nene Hilario           C  Washington Wizards  13.000000
Marcin Gortat          C  Washington Wizards  11.217391
DeJuan Blair           C  Washington Wizards   2.000000

[69 rows x 3 columns]

To filter rows for player who make 20 million or more:

In [132]:
nba_by_player.loc[nba_by_player["SALARY"] >= 20]

POSITION                   TEAM     SALARY
PLAYER                                                    
Joe Johnson           SF          Brooklyn Nets  24.894863
Derrick Rose          PG          Chicago Bulls  20.093064
LeBron James          SF    Cleveland Cavaliers  22.970500
Dwight Howard          C        Houston Rockets  22.359364
Chris Paul            PG   Los Angeles Clippers  21.468695
Kobe Bryant           SF     Los Angeles Lakers  25.000000
Chris Bosh            PF             Miami Heat  22.192730
Dwyane Wade           SG             Miami Heat  20.000000
Carmelo Anthony       SF        New York Knicks  22.875000
Kevin Durant          SF  Oklahoma City Thunder  20.158622

We can also use more than one condition, but the syntax is strange (when compared to "straight" Python, although similar to NumPy):

* Each condition must be surrounded by parentheses.
* We use `&` for `and`, `|` for `or`, and `~` for `not`.

For example, for if we want point guards (PG) that make over 12 million:

In [133]:
nba_by_player.loc[(nba_by_player["POSITION"] == "PG") & (nba_by_player["SALARY"] > 12)]

POSITION                    TEAM     SALARY
PLAYER                                                       
Derrick Rose            PG           Chicago Bulls  20.093064
Kyrie Irving            PG     Cleveland Cavaliers  16.407501
Reggie Jackson          PG         Detroit Pistons  13.913044
Ty Lawson               PG         Houston Rockets  12.404495
Chris Paul              PG    Los Angeles Clippers  21.468695
Goran Dragic            PG              Miami Heat  14.783000
Ricky Rubio             PG  Minnesota Timberwolves  12.700000
Russell Westbrook       PG   Oklahoma City Thunder  16.744218
Eric Bledsoe            PG            Phoenix Suns  13.500000
Tony Parker             PG       San Antonio Spurs  13.437500
John Wall               PG      Washington Wizards  15.851950

Here is a more complex (and contrived) example: suppose we want rows for players that either make less than 10 million, or play as a center, but not for the Chicago Bulls:

In [134]:
nba_by_player.loc[(nba_by_player["SALARY"] < 10) | 
                  ((~(nba_by_player["TEAM"] == "Chicago Bulls") & 
                    (nba_by_player["POSITION"] == "C"))
                  )]

POSITION                TEAM     SALARY
PLAYER                                                 
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Jeff Teague           PG       Atlanta Hawks   8.000000
Kyle Korver           SG       Atlanta Hawks   5.746479
Thabo Sefolosha       SF       Atlanta Hawks   4.000000
...                  ...                 ...        ...
Gary Neal             PG  Washington Wizards   2.139000
DeJuan Blair           C  Washington Wizards   2.000000
Kelly Oubre Jr.       SF  Washington Wizards   1.920240
Garrett Temple        SG  Washington Wizards   1.100602
Jarell Eddie          SG  Washington Wizards   0.561716

[363 rows x 3 columns]

The series methods [between](https://pandas.pydata.org/docs/reference/api/pandas.Series.between.html) and [isin](https://pandas.pydata.org/docs/reference/api/pandas.Series.isin.html) are also helpful when filtering.

For example, for salaries above and including 10 million, but below (and excluding) 15, we can do: 

In [135]:
nba_by_player.loc[nba_by_player["SALARY"].between(10, 15, inclusive="left")]

POSITION                    TEAM     SALARY
PLAYER                                                      
Al Horford              C           Atlanta Hawks  12.000000
Thaddeus Young         PF           Brooklyn Nets  11.235955
Al Jefferson            C       Charlotte Hornets  13.500000
Nicolas Batum          SG       Charlotte Hornets  13.125306
Kemba Walker           PG       Charlotte Hornets  12.000000
Joakim Noah             C           Chicago Bulls  13.400000
Tristan Thompson        C     Cleveland Cavaliers  14.260870
Brendan Haywood         C     Cleveland Cavaliers  10.522500
Danilo Gallinari       SF          Denver Nuggets  14.000000
Kenneth Faried         PF          Denver Nuggets  11.235955
Wilson Chandler        SF          Denver Nuggets  10.449438
Reggie Jackson         PG         Detroit Pistons  13.913044
Draymond Green         PF   Golden State Warriors  14.260870
Andrew Bogut            C   Golden State Warriors  13.800000
Andre Iguodala         SF   Golden State Warriors  11.710456
Stephen Curry          PG   Golden State Warriors  11.370786
Ty Lawson              PG         Houston Rockets  12.404495
Monta Ellis            SG          Indiana Pacers  10.300000
Goran Dragic           PG              Miami Heat  14.783000
Luol Deng              SF              Miami Heat  10.151612
Khris Middleton        SG         Milwaukee Bucks  14.700000
Ricky Rubio            PG  Minnesota Timberwolves  12.700000
Nikola Pekovic          C  Minnesota Timberwolves  12.100000
Tyreke Evans           SG    New Orleans Pelicans  10.734586
Jrue Holiday           PG    New Orleans Pelicans  10.595507
Robin Lopez             C         New York Knicks  12.650000
Serge Ibaka            PF   Oklahoma City Thunder  12.250000
Nikola Vucevic          C           Orlando Magic  11.250000
Gerald Wallace         SF      Philadelphia 76ers  10.105855
Eric Bledsoe           PG            Phoenix Suns  13.500000
Tyson Chandler          C            Phoenix Suns  13.000000
Rudy Gay               SF        Sacramento Kings  12.403101
Tony Parker            PG       San Antonio Spurs  13.437500
Danny Green            SG       San Antonio Spurs  10.000000
DeMarre Carroll        SF         Toronto Raptors  13.600000
Kyle Lowry             PG         Toronto Raptors  12.000000
DeMar DeRozan          SG         Toronto Raptors  10.050000
Nene Hilario            C      Washington Wizards  13.000000
Marcin Gortat           C      Washington Wizards  11.217391

The inclusion options for `.between` are:

* `inclusive="both"` includes both, the *default*;
* `inclusive="left"` includes left boundary only;
* `inclusive="right"` includes right boundary only;
* `inclusive="neither"` includes neither boundary.

If we want players that play for either the Chicago Bulls or New York Knicks:

In [136]:
nba_by_player.loc[nba_by_player["TEAM"].isin(["Chicago Bulls", "New York Knicks"])]

POSITION             TEAM     SALARY
PLAYER                                                     
Derrick Rose                 PG    Chicago Bulls  20.093064
Jimmy Butler                 SG    Chicago Bulls  16.407500
Joakim Noah                   C    Chicago Bulls  13.400000
Pau Gasol                     C    Chicago Bulls   7.448760
Nikola Mirotic               PF    Chicago Bulls   5.543725
Mike Dunleavy                SG    Chicago Bulls   4.500000
Kirk Hinrich                 SG    Chicago Bulls   2.854940
Doug McDermott               SF    Chicago Bulls   2.380440
Aaron Brooks                 PG    Chicago Bulls   2.250000
Tony Snell                   SF    Chicago Bulls   1.535880
Bobby Portis                 PF    Chicago Bulls   1.391160
E'Twaun Moore                SG    Chicago Bulls   1.015421
Carmelo Anthony              SF  New York Knicks  22.875000
Robin Lopez                   C  New York Knicks  12.650000
Arron Afflalo                SG  New York Knicks   8.000000
Jose Calderon                PG  New York Knicks   7.402812
Kristaps Porzingis           PF  New York Knicks   4.131720
Derrick Williams             PF  New York Knicks   4.000000
Kyle O'Quinn                 PF  New York Knicks   3.750000
Lance Thomas                 SF  New York Knicks   1.636842
Lou Amundson                 PF  New York Knicks   1.635476
Jerian Grant                 PG  New York Knicks   1.572360
Jimmer Fredette              PG  New York Knicks   0.874837
Langston Galloway            SG  New York Knicks   0.845059
Thanasis Antetokounmpo       SF  New York Knicks   0.030888

Another helpful series method for filtering is [str.contains](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).  For instance, let's find all the John's in `nba`:

In [137]:
nba.loc[nba["PLAYER"].str.contains("John")]

PLAYER POSITION                  TEAM     SALARY
29       Joe Johnson       SF         Brooklyn Nets  24.894863
92      John Jenkins       SG      Dallas Mavericks   0.981348
112  Stanley Johnson       SF       Detroit Pistons   2.841960
164   Wesley Johnson       SF  Los Angeles Clippers   1.100602
209    Tyler Johnson       SG            Miami Heat   0.845059
218      John Henson       PF       Milwaukee Bucks   2.943221
324  Orlando Johnson       SG          Phoenix Suns   0.055722
376    James Johnson       PF       Toronto Raptors   2.500000
394    Chris Johnson       SF             Utah Jazz   0.981348
400        John Wall       PG    Washington Wizards  15.851950

Note that the search is case sensitive, but it does get last names starting with `"John"`.  We could add a space to the string, in this case:

In [138]:
nba.loc[nba["PLAYER"].str.contains("John ")]

PLAYER POSITION                TEAM     SALARY
92   John Jenkins       SG    Dallas Mavericks   0.981348
218   John Henson       PF     Milwaukee Bucks   2.943221
400     John Wall       PG  Washington Wizards  15.851950

We can also do it with `nba_by_player`, but we have to use the index:

In [139]:
nba_by_player.loc[nba_by_player.index.str.contains("John ")]

POSITION                TEAM     SALARY
PLAYER                                              
John Jenkins       SG    Dallas Mavericks   0.981348
John Henson        PF     Milwaukee Bucks   2.943221
John Wall          PG  Washington Wizards  15.851950

### Filtering with `query`

An alternative to filtering with `loc` is to use [query](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html).

It's syntax is most often simpler, although somewhat strange.  It is somewhat less capable than `loc`, but it is often more convenient to use.

`query` takes a *string* as its argument, and instead of specifying the column from the data frame, we can use simply the column's label in our conditions.  Moreover, we can use the more usual Python syntax, including the use of `and`, `or`, and `not`.

Also note that `query` uses parentheses `( )` (as it is indeed a method), unlike `loc` (and `iloc`), which use square brackets `[ ]`.

To illustrate its use, we rework the examples done using `loc` above here, now using `query`.

To find the rows of centers:

In [140]:
nba_by_player.query("POSITION == 'C'")

POSITION                TEAM     SALARY
PLAYER                                                 
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Walter Tavares         C       Atlanta Hawks   1.000000
Tyler Zeller           C      Boston Celtics   2.616975
Jared Sullinger        C      Boston Celtics   2.569260
...                  ...                 ...        ...
Rudy Gobert            C           Utah Jazz   1.175880
Jeff Withey            C           Utah Jazz   0.947276
Nene Hilario           C  Washington Wizards  13.000000
Marcin Gortat          C  Washington Wizards  11.217391
DeJuan Blair           C  Washington Wizards   2.000000

[69 rows x 3 columns]

Note that we do need quotes for `C`, but not for `POSITION`, since it is a column label.

To find players who make 20 million or more:

In [141]:
nba_by_player.query("SALARY >= 20")

POSITION                   TEAM     SALARY
PLAYER                                                    
Joe Johnson           SF          Brooklyn Nets  24.894863
Derrick Rose          PG          Chicago Bulls  20.093064
LeBron James          SF    Cleveland Cavaliers  22.970500
Dwight Howard          C        Houston Rockets  22.359364
Chris Paul            PG   Los Angeles Clippers  21.468695
Kobe Bryant           SF     Los Angeles Lakers  25.000000
Chris Bosh            PF             Miami Heat  22.192730
Dwyane Wade           SG             Miami Heat  20.000000
Carmelo Anthony       SF        New York Knicks  22.875000
Kevin Durant          SF  Oklahoma City Thunder  20.158622

To find point guards that make over 12 million:

In [142]:
nba_by_player.query("POSITION == 'PG' and SALARY > 12")

POSITION                    TEAM     SALARY
PLAYER                                                       
Derrick Rose            PG           Chicago Bulls  20.093064
Kyrie Irving            PG     Cleveland Cavaliers  16.407501
Reggie Jackson          PG         Detroit Pistons  13.913044
Ty Lawson               PG         Houston Rockets  12.404495
Chris Paul              PG    Los Angeles Clippers  21.468695
Goran Dragic            PG              Miami Heat  14.783000
Ricky Rubio             PG  Minnesota Timberwolves  12.700000
Russell Westbrook       PG   Oklahoma City Thunder  16.744218
Eric Bledsoe            PG            Phoenix Suns  13.500000
Tony Parker             PG       San Antonio Spurs  13.437500
John Wall               PG      Washington Wizards  15.851950

To find rows for players that either make less than 10 million, or play as a center but not for the Chicago Bulls:

In [143]:
nba_by_player.query("SALARY < 10 or (not TEAM == 'Chicago Bulls' and POSITION == 'C')")

POSITION                TEAM     SALARY
PLAYER                                                 
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Jeff Teague           PG       Atlanta Hawks   8.000000
Kyle Korver           SG       Atlanta Hawks   5.746479
Thabo Sefolosha       SF       Atlanta Hawks   4.000000
...                  ...                 ...        ...
Gary Neal             PG  Washington Wizards   2.139000
DeJuan Blair           C  Washington Wizards   2.000000
Kelly Oubre Jr.       SF  Washington Wizards   1.920240
Garrett Temple        SG  Washington Wizards   1.100602
Jarell Eddie          SG  Washington Wizards   0.561716

[363 rows x 3 columns]

For salaries above and including 10 million, but below (and excluding) 15, we can do: 

In [144]:
nba_by_player.query("10 <= SALARY < 15")

POSITION                    TEAM     SALARY
PLAYER                                                      
Al Horford              C           Atlanta Hawks  12.000000
Thaddeus Young         PF           Brooklyn Nets  11.235955
Al Jefferson            C       Charlotte Hornets  13.500000
Nicolas Batum          SG       Charlotte Hornets  13.125306
Kemba Walker           PG       Charlotte Hornets  12.000000
Joakim Noah             C           Chicago Bulls  13.400000
Tristan Thompson        C     Cleveland Cavaliers  14.260870
Brendan Haywood         C     Cleveland Cavaliers  10.522500
Danilo Gallinari       SF          Denver Nuggets  14.000000
Kenneth Faried         PF          Denver Nuggets  11.235955
Wilson Chandler        SF          Denver Nuggets  10.449438
Reggie Jackson         PG         Detroit Pistons  13.913044
Draymond Green         PF   Golden State Warriors  14.260870
Andrew Bogut            C   Golden State Warriors  13.800000
Andre Iguodala         SF   Golden State Warriors  11.710456
Stephen Curry          PG   Golden State Warriors  11.370786
Ty Lawson              PG         Houston Rockets  12.404495
Monta Ellis            SG          Indiana Pacers  10.300000
Goran Dragic           PG              Miami Heat  14.783000
Luol Deng              SF              Miami Heat  10.151612
Khris Middleton        SG         Milwaukee Bucks  14.700000
Ricky Rubio            PG  Minnesota Timberwolves  12.700000
Nikola Pekovic          C  Minnesota Timberwolves  12.100000
Tyreke Evans           SG    New Orleans Pelicans  10.734586
Jrue Holiday           PG    New Orleans Pelicans  10.595507
Robin Lopez             C         New York Knicks  12.650000
Serge Ibaka            PF   Oklahoma City Thunder  12.250000
Nikola Vucevic          C           Orlando Magic  11.250000
Gerald Wallace         SF      Philadelphia 76ers  10.105855
Eric Bledsoe           PG            Phoenix Suns  13.500000
Tyson Chandler          C            Phoenix Suns  13.000000
Rudy Gay               SF        Sacramento Kings  12.403101
Tony Parker            PG       San Antonio Spurs  13.437500
Danny Green            SG       San Antonio Spurs  10.000000
DeMarre Carroll        SF         Toronto Raptors  13.600000
Kyle Lowry             PG         Toronto Raptors  12.000000
DeMar DeRozan          SG         Toronto Raptors  10.050000
Nene Hilario            C      Washington Wizards  13.000000
Marcin Gortat           C      Washington Wizards  11.217391

For players that play for either the Chicago Bulls or New York Knicks:

In [145]:
nba_by_player.query("TEAM in ['Chicago Bulls', 'New York Knicks']")

POSITION             TEAM     SALARY
PLAYER                                                     
Derrick Rose                 PG    Chicago Bulls  20.093064
Jimmy Butler                 SG    Chicago Bulls  16.407500
Joakim Noah                   C    Chicago Bulls  13.400000
Pau Gasol                     C    Chicago Bulls   7.448760
Nikola Mirotic               PF    Chicago Bulls   5.543725
Mike Dunleavy                SG    Chicago Bulls   4.500000
Kirk Hinrich                 SG    Chicago Bulls   2.854940
Doug McDermott               SF    Chicago Bulls   2.380440
Aaron Brooks                 PG    Chicago Bulls   2.250000
Tony Snell                   SF    Chicago Bulls   1.535880
Bobby Portis                 PF    Chicago Bulls   1.391160
E'Twaun Moore                SG    Chicago Bulls   1.015421
Carmelo Anthony              SF  New York Knicks  22.875000
Robin Lopez                   C  New York Knicks  12.650000
Arron Afflalo                SG  New York Knicks   8.000000
Jose Calderon                PG  New York Knicks   7.402812
Kristaps Porzingis           PF  New York Knicks   4.131720
Derrick Williams             PF  New York Knicks   4.000000
Kyle O'Quinn                 PF  New York Knicks   3.750000
Lance Thomas                 SF  New York Knicks   1.636842
Lou Amundson                 PF  New York Knicks   1.635476
Jerian Grant                 PG  New York Knicks   1.572360
Jimmer Fredette              PG  New York Knicks   0.874837
Langston Galloway            SG  New York Knicks   0.845059
Thanasis Antetokounmpo       SF  New York Knicks   0.030888

The last example, where we find players named John, is one in which we have the rare case that it is a lot more complex with `query` than with `loc`, so we skip it here.

#### Spaces in Column Labels with `query`

When using `query` we enter the column labels for filtering without quotes.  This poses a problem when there are spaces in the column label.  The solution is to surround the column label (inside the argument string) by back ticks `` ` ` ``.

Let's rename the SALARY by YEARLY SALARY (just so we can illustrate the method) and filter for values less than 5:

In [146]:
nba_by_player.rename(columns={"SALARY": "YEARLY SALARY"}).query("`YEARLY SALARY` < 5")

POSITION                TEAM  YEARLY SALARY
PLAYER                                                      
Thabo Sefolosha        SF       Atlanta Hawks       4.000000
Mike Scott             PF       Atlanta Hawks       3.333333
Kent Bazemore          SF       Atlanta Hawks       2.000000
Dennis Schroder        PG       Atlanta Hawks       1.763400
Tim Hardaway Jr.       SG       Atlanta Hawks       1.304520
...                   ...                 ...            ...
Gary Neal              PG  Washington Wizards       2.139000
DeJuan Blair            C  Washington Wizards       2.000000
Kelly Oubre Jr.        SF  Washington Wizards       1.920240
Garrett Temple         SG  Washington Wizards       1.100602
Jarell Eddie           SG  Washington Wizards       0.561716

[270 rows x 3 columns]

#### Variables with `query`

We cannot use values stored in a variable directly with `query`, since its argument is a string.  (It works fine with `loc`.)  But, we can do it by adding a `@` before the variable name in the argument string:

In [147]:
lower_bound_salary = 18  # you can change this value, without changing the next line

nba_by_player.query("SALARY >= @lower_bound_salary")

POSITION                   TEAM     SALARY
PLAYER                                                      
Paul Millsap            PF          Atlanta Hawks  18.671659
Joe Johnson             SF          Brooklyn Nets  24.894863
Derrick Rose            PG          Chicago Bulls  20.093064
LeBron James            SF    Cleveland Cavaliers  22.970500
Kevin Love              PF    Cleveland Cavaliers  19.689000
Dwight Howard            C        Houston Rockets  22.359364
Chris Paul              PG   Los Angeles Clippers  21.468695
Blake Griffin           PF   Los Angeles Clippers  18.907726
Kobe Bryant             SF     Los Angeles Lakers  25.000000
Marc Gasol               C      Memphis Grizzlies  19.688000
Chris Bosh              PF             Miami Heat  22.192730
Dwyane Wade             SG             Miami Heat  20.000000
Carmelo Anthony         SF        New York Knicks  22.875000
Kevin Durant            SF  Oklahoma City Thunder  20.158622
LaMarcus Aldridge       PF      San Antonio Spurs  19.689000

## Extracting Single Entry

We can extract a single entry with either `loc`, using labels, or with `iloc`, using positions:

In [148]:
nba

PLAYER POSITION                TEAM     SALARY
0       Paul Millsap       PF       Atlanta Hawks  18.671659
1         Al Horford        C       Atlanta Hawks  12.000000
2     Tiago Splitter        C       Atlanta Hawks   9.756250
3        Jeff Teague       PG       Atlanta Hawks   8.000000
4        Kyle Korver       SG       Atlanta Hawks   5.746479
..               ...      ...                 ...        ...
412        Gary Neal       PG  Washington Wizards   2.139000
413     DeJuan Blair        C  Washington Wizards   2.000000
414  Kelly Oubre Jr.       SF  Washington Wizards   1.920240
415   Garrett Temple       SG  Washington Wizards   1.100602
416     Jarell Eddie       SG  Washington Wizards   0.561716

[417 rows x 4 columns]

In [149]:
nba_by_player.loc["Jeff Teague", "TEAM"]

'Atlanta Hawks'

In [150]:
nba_by_player.iloc[3, 1]

'Atlanta Hawks'

Note that if we don't assign a column as the index, the index is numerical:

In [151]:
nba.loc[3, "TEAM"]

'Atlanta Hawks'

## Dropping Columns

We can use the method `drop` to remove columns.  We pass the column labels as the value for `columns=`:

In [152]:
nba_by_player.drop(columns="POSITION")

TEAM     SALARY
PLAYER                                        
Paul Millsap          Atlanta Hawks  18.671659
Al Horford            Atlanta Hawks  12.000000
Tiago Splitter        Atlanta Hawks   9.756250
Jeff Teague           Atlanta Hawks   8.000000
Kyle Korver           Atlanta Hawks   5.746479
...                             ...        ...
Gary Neal        Washington Wizards   2.139000
DeJuan Blair     Washington Wizards   2.000000
Kelly Oubre Jr.  Washington Wizards   1.920240
Garrett Temple   Washington Wizards   1.100602
Jarell Eddie     Washington Wizards   0.561716

[417 rows x 2 columns]

We can also drop more than one columns by passing a list:

In [153]:
nba_by_player.drop(columns=["POSITION", "TEAM"])

SALARY
PLAYER                    
Paul Millsap     18.671659
Al Horford       12.000000
Tiago Splitter    9.756250
Jeff Teague       8.000000
Kyle Korver       5.746479
...                    ...
Gary Neal         2.139000
DeJuan Blair      2.000000
Kelly Oubre Jr.   1.920240
Garrett Temple    1.100602
Jarell Eddie      0.561716

[417 rows x 1 columns]

## Adding Columns

The recommended way to add a column is the use the [assign method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.assign.html).

For instance, let's add an AGE columns to our `nba_by_player` data frame.  Here we will use fake, randomly generated ages.

In [154]:
ages = np.random.randint(18, 41, len(nba_by_player))  # random ages from 18 to 40

To add the array `ages` as the `AGE` columns, we do:

In [155]:
nba_by_player.assign(AGE=ages)

POSITION                TEAM     SALARY  AGE
PLAYER                                                      
Paul Millsap          PF       Atlanta Hawks  18.671659   24
Al Horford             C       Atlanta Hawks  12.000000   32
Tiago Splitter         C       Atlanta Hawks   9.756250   37
Jeff Teague           PG       Atlanta Hawks   8.000000   39
Kyle Korver           SG       Atlanta Hawks   5.746479   18
...                  ...                 ...        ...  ...
Gary Neal             PG  Washington Wizards   2.139000   33
DeJuan Blair           C  Washington Wizards   2.000000   24
Kelly Oubre Jr.       SF  Washington Wizards   1.920240   33
Garrett Temple        SG  Washington Wizards   1.100602   38
Jarell Eddie          SG  Washington Wizards   0.561716   39

[417 rows x 4 columns]

Note that we use the new column label as a *variable name*.

Also, as all methods we've seen so far, `assign` returns a **new data frame, and does not change the original**:

In [156]:
nba_by_player

POSITION                TEAM     SALARY
PLAYER                                                 
Paul Millsap          PF       Atlanta Hawks  18.671659
Al Horford             C       Atlanta Hawks  12.000000
Tiago Splitter         C       Atlanta Hawks   9.756250
Jeff Teague           PG       Atlanta Hawks   8.000000
Kyle Korver           SG       Atlanta Hawks   5.746479
...                  ...                 ...        ...
Gary Neal             PG  Washington Wizards   2.139000
DeJuan Blair           C  Washington Wizards   2.000000
Kelly Oubre Jr.       SF  Washington Wizards   1.920240
Garrett Temple        SG  Washington Wizards   1.100602
Jarell Eddie          SG  Washington Wizards   0.561716

[417 rows x 3 columns]

Of course, if we want to change the original, we can overwrite it as usual with

```python
nba_by_player = nba_by_player.assign(AGE=ages)
```

An alternative that adds the columns directly into the data frame (effectively changing it) is

```python
nba_by_player["AGE"] = ages
```

(Again, this last one *does* change the original `nba_by_player`!)

We can also create more than one column with `assign`, by separating the assignments by commas:

In [157]:
heights = np.random.randint(72, 90, len(nba_by_player))  # random heights between 6' and 7'5"

nba_by_player.assign(AGE=ages, HEIGHT=heights)

POSITION                TEAM     SALARY  AGE  HEIGHT
PLAYER                                                              
Paul Millsap          PF       Atlanta Hawks  18.671659   24      87
Al Horford             C       Atlanta Hawks  12.000000   32      83
Tiago Splitter         C       Atlanta Hawks   9.756250   37      82
Jeff Teague           PG       Atlanta Hawks   8.000000   39      79
Kyle Korver           SG       Atlanta Hawks   5.746479   18      72
...                  ...                 ...        ...  ...     ...
Gary Neal             PG  Washington Wizards   2.139000   33      79
DeJuan Blair           C  Washington Wizards   2.000000   24      72
Kelly Oubre Jr.       SF  Washington Wizards   1.920240   33      82
Garrett Temple        SG  Washington Wizards   1.100602   38      76
Jarell Eddie          SG  Washington Wizards   0.561716   39      88

[417 rows x 5 columns]

Often we need to add a column added from some computation, but it is basically the same.  For instance, let's add a column "Difference from Average", that has the difference between players salary and the average salary from the data frame.

A first obstacle here is that we have space in the column label, while `assign` uses a variable name, which cannot contain spaces.  So, we need to use a temporary label for the column and then rename it:

In [158]:
average_salary = nba_by_player["SALARY"].mean()

nba_by_player.assign(t=nba_by_player["SALARY"] - average_salary).rename(columns={"t": "Difference from Avarage"})

POSITION                TEAM     SALARY  \
PLAYER                                                    
Paul Millsap          PF       Atlanta Hawks  18.671659   
Al Horford             C       Atlanta Hawks  12.000000   
Tiago Splitter         C       Atlanta Hawks   9.756250   
Jeff Teague           PG       Atlanta Hawks   8.000000   
Kyle Korver           SG       Atlanta Hawks   5.746479   
...                  ...                 ...        ...   
Gary Neal             PG  Washington Wizards   2.139000   
DeJuan Blair           C  Washington Wizards   2.000000   
Kelly Oubre Jr.       SF  Washington Wizards   1.920240   
Garrett Temple        SG  Washington Wizards   1.100602   
Jarell Eddie          SG  Washington Wizards   0.561716   

                 Difference from Avarage  
PLAYER                                    
Paul Millsap                   13.596845  
Al Horford                      6.925186  
Tiago Splitter                  4.681436  
Jeff Teague                     2.925186  
Kyle Korver                     0.671665  
...                                  ...  
Gary Neal                      -2.935814  
DeJuan Blair                   -3.074814  
Kelly Oubre Jr.                -3.154574  
Garrett Temple                 -3.974212  
Jarell Eddie                   -4.513098  

[417 rows x 4 columns]

## Columns Statistics

We can get statistics for numerical columns with `.describe`:

In [159]:
nba_by_player.assign(AGE=ages, HEIGHT=heights).describe()

SALARY         AGE      HEIGHT
count  417.000000  417.000000  417.000000
mean     5.074814   28.342926   80.721823
std      5.221437    6.642658    5.132788
min      0.030888   18.000000   72.000000
25%      1.270964   22.000000   76.000000
50%      3.000000   28.000000   81.000000
75%      7.000000   34.000000   85.000000
max     25.000000   40.000000   89.000000

The percentages are the *quartiles*.  For instance, since the value of SALARY in the `25%` row is (about) $1.27$, this means that a quarter of the players make $1.27$ millions or less.   Since the value of SALARY in the `75%` row is $7$, this means that three quarters of the players make $7$ millions or less.

`std` stands for *standard deviation*.

## Columns Values

For categorical columns, we can find which values are present by selecting it and using the `unique` method.

So, if we want to see the list of teams in our data frame:

In [160]:
nba_by_player["TEAM"].unique()

array(['Atlanta Hawks', 'Boston Celtics', 'Brooklyn Nets',
       'Charlotte Hornets', 'Chicago Bulls', 'Cleveland Cavaliers',
       'Dallas Mavericks', 'Denver Nuggets', 'Detroit Pistons',
       'Golden State Warriors', 'Houston Rockets', 'Indiana Pacers',
       'Los Angeles Clippers', 'Los Angeles Lakers', 'Memphis Grizzlies',
       'Miami Heat', 'Milwaukee Bucks', 'Minnesota Timberwolves',
       'New Orleans Pelicans', 'New York Knicks', 'Oklahoma City Thunder',
       'Orlando Magic', 'Philadelphia 76ers', 'Phoenix Suns',
       'Portland Trail Blazers', 'Sacramento Kings', 'San Antonio Spurs',
       'Toronto Raptors', 'Utah Jazz', 'Washington Wizards'], dtype=object)

## `groupby`

Suppose now we want to extra some more "indirect" information about the data.  For instance, suppose we would like to know what is the average salary paid *per team*.

So, we want some information about a certain *category* in the data frame (like, team or position).  The tool for this sort of computation in pandas is [groupby](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html).  It basically groups the data by the given category.

When we group by team, in our example, we will have various values for the salary, and we need to tell pandas how to *aggregate* all these salaries for a single team.  In our case, we want the average, so we use the `mean` groupby method.

On the other hand, since we cannot compute the average of the different values for the position, we need to first select only the columns we need, i.e., TEAM and SALARY in this case:

In [161]:
nba_by_player[["TEAM", "SALARY"]].groupby("TEAM").mean()

SALARY
TEAM                             
Atlanta Hawks            4.969507
Boston Celtics           3.352367
Brooklyn Nets            4.408229
Charlotte Hornets        4.672355
Chicago Bulls            6.568407
Cleveland Cavaliers     10.231241
Dallas Mavericks         5.978414
Denver Nuggets           4.459243
Detroit Pistons          4.221176
Golden State Warriors    6.720367
Houston Rockets          7.107153
Indiana Pacers           4.822694
Los Angeles Clippers     5.082624
Los Angeles Lakers       6.237086
Memphis Grizzlies        4.466497
Miami Heat               6.794056
Milwaukee Bucks          4.019873
Minnesota Timberwolves   5.065186
New Orleans Pelicans     5.032163
New York Knicks          5.338846
Oklahoma City Thunder    6.052010
Orlando Magic            5.544567
Philadelphia 76ers       3.267796
Phoenix Suns             2.971813
Portland Trail Blazers   3.246206
Sacramento Kings         6.216808
San Antonio Spurs        6.511698
Toronto Raptors          4.392507
Utah Jazz                3.095993
Washington Wizards       5.296912

Note that the category we used to group by becomes the index of the resulting data frame.  

If we have more columns from which we can use the same aggregating function (the average in this case), we can pass them all:

In [162]:
nba_by_player.assign(AGE=ages, HEIGHT=heights).drop(columns="POSITION").groupby("TEAM").mean()

SALARY        AGE     HEIGHT
TEAM                                                   
Atlanta Hawks            4.969507  29.928571  81.642857
Boston Celtics           3.352367  28.733333  80.800000
Brooklyn Nets            4.408229  30.692308  82.153846
Charlotte Hornets        4.672355  27.333333  81.611111
Chicago Bulls            6.568407  26.916667  82.916667
Cleveland Cavaliers     10.231241  26.500000  79.400000
Dallas Mavericks         5.978414  27.454545  79.818182
Denver Nuggets           4.459243  27.142857  83.357143
Detroit Pistons          4.221176  28.900000  81.200000
Golden State Warriors    6.720367  23.357143  82.357143
Houston Rockets          7.107153  29.416667  77.833333
Indiana Pacers           4.822694  29.000000  78.384615
Los Angeles Clippers     5.082624  28.384615  79.384615
Los Angeles Lakers       6.237086  28.090909  81.727273
Memphis Grizzlies        4.466497  28.761905  81.714286
Miami Heat               6.794056  26.000000  80.083333
Milwaukee Bucks          4.019873  28.461538  83.153846
Minnesota Timberwolves   5.065186  29.153846  79.384615
New Orleans Pelicans     5.032163  28.312500  82.437500
New York Knicks          5.338846  27.692308  78.384615
Oklahoma City Thunder    6.052010  28.375000  81.250000
Orlando Magic            5.544567  30.214286  80.000000
Philadelphia 76ers       3.267796  27.076923  81.461538
Phoenix Suns             2.971813  28.823529  80.882353
Portland Trail Blazers   3.246206  26.928571  79.500000
Sacramento Kings         6.216808  30.272727  80.454545
San Antonio Spurs        6.511698  27.846154  80.538462
Toronto Raptors          4.392507  27.764706  78.176471
Utah Jazz                3.095993  31.058824  79.764706
Washington Wizards       5.296912  30.294118  80.823529

Similarly, if we want the *median* salary per *position*:

In [163]:
nba_by_player[["POSITION", "SALARY"]].groupby("POSITION").median()

SALARY
POSITION          
C         4.204200
PF        3.300000
PG        3.036927
SF        2.658953
SG        2.233779

If we simply want to count how many rows we have for each category, we use the aggregating groupby method `size`.  In this case, there is no need to select/drop columns.

For instance, to see how many players we have in each position:

In [164]:
nba_by_player.groupby("POSITION").size()

POSITION
C     69
PF    85
PG    85
SF    82
SG    96
dtype: int64

Note that the result is a *series*, and not a data frame.  We can make it into a data frame using the method `to_frame`:

In [165]:
nba_by_player.groupby("POSITION").size().to_frame()

0
POSITION    
C         69
PF        85
PG        85
SF        82
SG        96

Note that the label for the column is just the number 0.  We can specify the label by passing is (as a string) to the method:

In [166]:
nba_by_player.groupby("POSITION").size().to_frame("Count")

Count
POSITION       
C            69
PF           85
PG           85
SF           82
SG           96

We can also use more than one aggregating function using the groupby method `agg`.  We pass a list of the aggregating methods we want to use *as strings*.  

For instance, to see the minimum and maximum salary paid by a team:

In [167]:
nba_by_player[["TEAM", "SALARY"]].groupby("TEAM").agg(["min", "max"])

SALARY           
                             min        max
TEAM                                       
Atlanta Hawks           0.525093  18.671659
Boston Celtics          1.148640   7.730337
Brooklyn Nets           0.845059  24.894863
Charlotte Hornets       0.525093  13.500000
Chicago Bulls           1.015421  20.093064
Cleveland Cavaliers     1.147276  22.970500
Dallas Mavericks        0.981348  15.493680
Denver Nuggets          0.099418  14.000000
Detroit Pistons         0.600000  13.913044
Golden State Warriors   0.289755  15.501000
Houston Rockets         1.000000  22.359364
Indiana Pacers          1.007026  17.120106
Los Angeles Clippers    0.111444  21.468695
Los Angeles Lakers      0.700000  25.000000
Memphis Grizzlies       0.055722  19.688000
Miami Heat              0.525093  22.192730
Milwaukee Bucks         0.855000  14.700000
Minnesota Timberwolves  1.149500  12.700000
New Orleans Pelicans    0.169883  15.514031
New York Knicks         0.030888  22.875000
Oklahoma City Thunder   0.289755  20.158622
Orlando Magic           0.061776  16.000000
Philadelphia 76ers      0.258489  10.105855
Phoenix Suns            0.049709  13.500000
Portland Trail Blazers  0.525093   8.042895
Sacramento Kings        1.015421  15.851950
San Antonio Spurs       0.947276  19.689000
Toronto Raptors         0.525093  13.600000
Utah Jazz               0.099418  15.409570
Washington Wizards      0.561716  15.851950

We can also group by more than one category, by passing a list as the argument.

For instance, to see the average salary per team and position:

In [168]:
nba_by_player.groupby(["TEAM", "POSITION"]).mean()

SALARY
TEAM               POSITION           
Atlanta Hawks      C          7.585417
                   PF        11.002496
                   PG         4.881700
                   SF         3.000000
                   SG         1.809692
...                                ...
Washington Wizards C          8.739130
                   PF         5.246667
                   PG         6.720472
                   SF         4.142925
                   SG         2.839248

[147 rows x 1 columns]

The order matters:

In [169]:
nba_by_player.groupby(["POSITION", "TEAM"]).mean()

SALARY
POSITION TEAM                         
C        Atlanta Hawks        7.585417
         Boston Celtics       2.450465
         Brooklyn Nets        1.362897
         Charlotte Hornets    6.772240
         Chicago Bulls       10.424380
...                                ...
SG       Sacramento Kings     3.410876
         San Antonio Spurs    6.407000
         Toronto Raptors      5.350000
         Utah Jazz            3.779006
         Washington Wizards   2.839248

[147 rows x 1 columns]

## Pivot Tables

Grouping by more than one category works well, but it can be a bit harder to visualize.  And alternative is to use *pivot tables*.  In this case, one of the two categories becomes the index (for rows) and the other becomes the columns.

Here is how we would use it, similar to the previous example, so we can visualize the average salary per team and position:

In [170]:
nba_by_player.pivot_table(index="TEAM", columns="POSITION", values="SALARY", aggfunc=np.mean)

POSITION                        C         PF         PG         SF        SG
TEAM                                                                        
Atlanta Hawks            7.585417  11.002496   4.881700   3.000000  1.809692
Boston Celtics           2.450465   3.085480   4.974652   4.417161  2.007554
Brooklyn Nets            1.362897   4.452514   3.900000  13.040323  1.741178
Charlotte Hornets        6.772240   4.685770   4.485304   3.766422  4.042384
Chicago Bulls           10.424380   3.467443  11.171532   1.958160  6.194465
Cleveland Cavaliers      7.752343  19.689000   6.551592  22.970500  8.988765
Dallas Mavericks         3.235482  11.913507   4.418183  15.361500  1.215174
Denver Nuggets           2.616305   7.024978   3.723620   7.195773  0.841949
Detroit Pistons          4.090697        NaN  13.913044   1.716216  4.580880
Golden State Warriors    6.541249   7.186367   8.457256   4.496695  9.000500
Houston Rockets         22.359364   1.755114   9.445491   4.919715  9.058536
Indiana Pacers           4.000000   2.471595   4.003513   6.968303  5.700000
Los Angeles Clippers     1.100602   9.509585   8.509072   2.238301  3.190104
Los Angeles Lakers       8.286783   2.253023   5.103120  10.306390  7.000000
Memphis Grizzlies        9.899722   3.364087   4.225988   5.709390  2.676720
Miami Heat                    NaN  10.197132   7.654046   3.564757  7.123384
Milwaukee Bucks          2.109294   4.047830   3.415350   1.319653  8.144347
Minnesota Timberwolves   6.426013   4.796280   5.376813   1.603210  6.421840
New Orleans Pelicans     4.850122   6.140243   3.830596   1.934208  8.806167
New York Knicks         12.650000   3.379299   3.283336   8.180910  4.422529
Oklahoma City Thunder    9.343270   5.821013   7.255246   6.522154  3.207057
Orlando Magic           11.250000   5.389060   3.051608  16.000000  3.740735
Philadelphia 76ers       4.604820   3.467184   2.144772   3.983638  1.769094
Phoenix Suns             8.403560   2.194903   3.592017   3.770540  0.869973
Portland Trail Blazers   4.470774   2.079564   4.236287   5.468477  2.124524
Sacramento Kings         8.983410        NaN   7.256780   5.562825  3.410876
San Antonio Spurs        4.650000  10.594093   5.987908   6.245388  6.407000
Toronto Raptors          3.105494   3.048442   4.756891   6.225972  5.350000
Utah Jazz                1.674385   2.654025   1.973742   6.146973  3.779006
Washington Wizards       8.739130   5.246667   6.720472   4.142925  2.839248

We can have more than a single value to aggregate:

In [171]:
nba_by_player.assign(AGE=ages, HEIGHT=heights).pivot_table(index="TEAM", columns="POSITION", values=["SALARY", "AGE"], aggfunc=np.mean)

AGE                                              \
POSITION                        C         PF         PG         SF         SG   
TEAM                                                                            
Atlanta Hawks           29.666667  27.500000  36.000000  23.500000  31.200000   
Boston Celtics          27.000000  27.500000  28.500000  32.000000  29.250000   
Brooklyn Nets           39.000000  32.333333  23.500000  30.000000  31.200000   
Charlotte Hornets       26.000000  26.666667  27.500000  28.500000  27.833333   
Chicago Bulls           30.500000  29.000000  30.500000  27.000000  22.250000   
Cleveland Cavaliers     28.500000  21.000000  27.666667  26.000000  21.000000   
Dallas Mavericks        30.000000  28.500000  25.750000  30.000000  26.000000   
Denver Nuggets          30.000000  25.000000  20.000000  31.500000  22.000000   
Detroit Pistons         27.666667        NaN  21.000000  29.750000  33.000000   
Golden State Warriors   27.000000  20.000000  23.500000  22.250000  25.000000   
Houston Rockets         18.000000  30.500000  35.000000  22.500000  32.666667   
Indiana Pacers          36.000000  34.666667  28.000000  25.250000  22.000000   
Los Angeles Clippers    23.000000  24.500000  28.333333  26.500000  31.800000   
Los Angeles Lakers      29.500000  25.750000  39.000000  30.000000  18.000000   
Memphis Grizzlies       28.000000  27.571429  28.750000  30.500000  29.500000   
Miami Heat                    NaN  26.666667  30.500000  23.500000  25.666667   
Milwaukee Bucks         20.000000  26.500000  32.750000  27.666667  27.666667   
Minnesota Timberwolves  25.333333  30.333333  32.000000  25.500000  32.500000   
New Orleans Pelicans    31.000000  27.666667  32.500000  24.333333  24.666667   
New York Knicks         25.000000  23.000000  27.000000  36.000000  27.000000   
Oklahoma City Thunder   30.000000  34.666667  30.666667  23.750000  25.750000   
Orlando Magic           33.000000  28.000000  32.000000  39.000000  27.666667   
Philadelphia 76ers      29.000000  30.750000  22.000000  23.333333  26.333333   
Phoenix Suns            30.000000  31.333333  28.500000  21.500000  29.833333   
Portland Trail Blazers  25.666667  26.000000  28.000000  29.000000  27.200000   
Sacramento Kings        34.666667        NaN  29.000000  32.666667  24.333333   
San Antonio Spurs       28.666667  28.000000  36.333333  23.333333  20.500000   
Toronto Raptors         24.333333  30.750000  26.600000  29.333333  27.500000   
Utah Jazz               27.000000  34.000000  32.400000  29.000000  32.000000   
Washington Wizards      26.000000  32.333333  30.666667  31.250000  30.750000   

                           SALARY                                             
POSITION                        C         PF         PG         SF        SG  
TEAM                                                                          
Atlanta Hawks            7.585417  11.002496   4.881700   3.000000  1.809692  
Boston Celtics           2.450465   3.085480   4.974652   4.417161  2.007554  
Brooklyn Nets            1.362897   4.452514   3.900000  13.040323  1.741178  
Charlotte Hornets        6.772240   4.685770   4.485304   3.766422  4.042384  
Chicago Bulls           10.424380   3.467443  11.171532   1.958160  6.194465  
Cleveland Cavaliers      7.752343  19.689000   6.551592  22.970500  8.988765  
Dallas Mavericks         3.235482  11.913507   4.418183  15.361500  1.215174  
Denver Nuggets           2.616305   7.024978   3.723620   7.195773  0.841949  
Detroit Pistons          4.090697        NaN  13.913044   1.716216  4.580880  
Golden State Warriors    6.541249   7.186367   8.457256   4.496695  9.000500  
Houston Rockets         22.359364   1.755114   9.445491   4.919715  9.058536  
Indiana Pacers           4.000000   2.471595   4.003513   6.968303  5.700000  
Los Angeles Clippers     1.100602   9.509585   8.509072   2.238301  3.190104  
Los Angeles Lakers       8.286783   2.253023   5.103120  10.306390  7.000000  
Memphis Grizz

Note the labels over the columns, specifying the values that were aggregated.

Note also that we have some strange entries above: `NaN`.  It stands for "not a number", and corresponds to entries that are not available.  For instance, the `NaN` in the Miami Heat row and C columns indicates that there are no centers in the Miami Heat team:

In [172]:
nba_by_player.query("TEAM == 'Miami Heat'")

POSITION        TEAM     SALARY
PLAYER                                          
Chris Bosh             PF  Miami Heat  22.192730
Dwyane Wade            SG  Miami Heat  20.000000
Goran Dragic           PG  Miami Heat  14.783000
Luol Deng              SF  Miami Heat  10.151612
Josh McRoberts         PF  Miami Heat   5.543725
Udonis Haslem          PF  Miami Heat   2.854940
Justise Winslow        SF  Miami Heat   2.481720
Henry Walker           SF  Miami Heat   1.100602
Tyler Johnson          SG  Miami Heat   0.845059
Greg Whittington       SF  Miami Heat   0.525093
Josh Richardson        SG  Miami Heat   0.525093
Corey Hawkins          PG  Miami Heat   0.525093

## Chaining

Often we want to perform a series of operations on a data frame.  This is often done by *chaining* the methods we need.  (We've already seen some examples above!)

Let's do one more example, which is again contrived, but illustrates the idea.  Suppose we want an array with the five teams that pay their players under 30 the most on average, in order of paying the best to worst.  (Note that we need to manually add the AGE column.)

Here is how it can be done:

In [173]:
nba_by_player.assign(AGE=ages).query("AGE < 30")[["TEAM", "SALARY"]].groupby("TEAM").mean().sort_values("SALARY", ascending=False).index.to_numpy()[:5]

array(['Cleveland Cavaliers', 'Los Angeles Lakers', 'Miami Heat',
       'Oklahoma City Thunder', 'Golden State Warriors'], dtype=object)

Note that the line is too long.  Unfortunately, since white space matters in Python, we cannot just break lines.  On the other hand, we can if we surround the command by parentheses:

In [174]:
(
    nba_by_player.assign(AGE=ages)  # add age column
    .query("AGE < 30")  # select under 30
    [["TEAM", "SALARY"]]  # select desired columns
    .groupby("TEAM").mean()  # groupby by team and take average
    .sort_values("SALARY", ascending=False)  # sort
    .index  # extract the index
    .to_numpy()  # make it into an array
    [:5]  # first 5
)

array(['Cleveland Cavaliers', 'Los Angeles Lakers', 'Miami Heat',
       'Oklahoma City Thunder', 'Golden State Warriors'], dtype=object)

Note that breaking the lines allows us to add comments as well! 

## Comments, Suggestions, Corrections

Please send your comments, suggestions, and corrections to lfinotti@utk.edu.